### ISSUE TITLE
Observable Asynchronous

### ISSUE THREAD

**Issue by pradhakrishnan** [ReactiveX/RxJava#2485]
Hi Team,

I have an issue with observable async operation. Basically I want to do some action in parallel and finally i need to zip all the observable results into single results. I noticed that all are going in the same thread so call goes synchronously. Can someone please help me to fix that issue.

E.g:

My Code
```
Observable<Set<Result>> rSet = Observable.zip(Observable.create(new ItemsOnSubscribe(data1),
Observable.create(new ItemsOnSubscribe(data2))
        new Func2<Set<Result>, Set<Result>, Set<Result>>() {

          @Override
          public Set<EntitledItem> call(Set<EntitledItem> t1,
              Set<EntitledItem> t2, ) {
            returnSet.addAll(t1);
            returnSet.addAll(t2);
            return returnSet;
          }
    });
```
My expectation here is call to Observable.create(new ItemsOnSubscribe(data1) and Observable.create(new ItemsOnSubscribe(data1) execute in parallel but it is going in sequentially.

Someone please help me on this.

Regards
Pranesh

**Comment by akarnokd** [comment]
Hello. You need to call ```subscribeOn()``` after each Observable has been created:

```java
Observable.zip(Observable.create(
    new ItemsOnSubscribe(data1)).subscribeOn(Schedulers.computation()),
    new ItemsOnSubscribe(data2)).subscribeOn(Schedulers.computation()),
...
```

If you do something blocking in ```ItemsOnSubscribe```, you are better off with ```Schedulers.io()```.

**Comment by pradhakrishnan** [comment]
Thanks Akarnokd, Really appreciate your help, Let me try the fix you told and will update the status.

**Comment by pradhakrishnan** [comment]
It works. Thanks Akarnokd.

**Comment by pradhakrishnan** [comment]
Hi Akarnokd,
I have another issue on this, before i subscribe the value my method returns empty set in response.

```
Observable.zip(Observable.create(new ItemsOnSubscribe(data1).subscribeOn(Schedulers.computation()),
Observable.create(new ItemsOnSubscribe(data2)).subscribeOn(Schedulers.computation()),
        new Func2<Set<Result>, Set<Result>, Set<Result>>() {

          @Override
          public Set<EntitledItem> call(Set<EntitledItem> t1,
              Set<EntitledItem> t2, ) {
            returnSet.addAll(t1);
            returnSet.addAll(t2);
            return returnSet;
          }
    }).subscribeOn(Schedulers.io()).subscribe(new Action1<Set<EntitledItem>>() {

				@Override
				public void call(Set<EntitledItem> t1) {
					response.getResponseObject().addAll(t1);
				}
			}, new Action1<Throwable>() {

				@Override
				public void call(Throwable t1) {
					logger.warn("************** Error occured ************ ");
					
				}
			});
```
Issue here is before it get subscribed my method returns empty set in reponse object.
Please help me to fix that.

**Comment by abersnaze** [comment]
By `my method returns` do you mean that the method that contains this code isn't waiting for the completion of the subscription?

It depends on what type server you are using.  If you are using something like servlet where the request thread has to block until all the work is done you should looking into using `.toBlocking()`.  This can be an anti-pattern to use more than once.  Ideally for blocking servers you should only have ONE call to `.toBlocking()` in the whole execution of that request.

If you are using an async server like netty (you might want to take a look at [RxNetty](https://github.com/ReactiveX/RxNetty)) or the async servlet spec then there should be some way letting the request know when the `onCompleted` happens that the request is done without blocking a thread.

**Comment by pradhakrishnan** [comment]
Thanks Abersnaze, I am using resin server so I think i have to use toBlocking() method.
